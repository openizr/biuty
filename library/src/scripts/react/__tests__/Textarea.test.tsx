/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import UITextarea from 'scripts/react/Textarea';
import { act, Simulate } from 'react-dom/test-utils';
import { render, unmountComponentAtNode } from 'react-dom';

type Target = any; // eslint-disable-line @typescript-eslint/no-explicit-any
let container = document.createElement('div');
jest.useFakeTimers();
jest.mock('scripts/helpers/markdown');
jest.mock('scripts/helpers/generateRandomId');

describe('react/UITextarea', () => {
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    jest.clearAllMocks();
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    ((container as unknown) as null) = null;
  });

  test('renders correctly - basic', () => {
    act(() => {
      render(<UITextarea name="test" modifiers="large" />, container);
    });
    const textarea = document.querySelector('textarea') as HTMLTextAreaElement;
    act(() => {
      Simulate.focus(textarea);
    });
    act(() => {
      Simulate.change(textarea, { target: { value: 'test' } as Target });
    });
    act(() => {
      Simulate.blur(textarea);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with id', () => {
    act(() => {
      render(<UITextarea name="test" id="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with cols and rows', () => {
    act(() => {
      render(<UITextarea name="test" cols={10} rows={50} />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with placeholder', () => {
    act(() => {
      render(<UITextarea name="test" placeholder="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with helper', () => {
    act(() => {
      render(<UITextarea name="test" helper="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with label', () => {
    act(() => {
      render(<UITextarea name="test" label="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with value', () => {
    act(() => {
      render(<UITextarea name="test" value="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with maxlength', () => {
    act(() => {
      render(<UITextarea name="test" maxlength={10} />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - disabled', () => {
    act(() => {
      render(<UITextarea name="test" modifiers="disabled" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - autocomplete off', () => {
    act(() => {
      render(<UITextarea name="test" autocomplete="off" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - readonly', () => {
    const onChange = jest.fn();
    act(() => {
      render(<UITextarea name="test" readonly onChange={onChange} />, container);
    });
    const textarea = document.querySelector('textarea') as HTMLTextAreaElement;
    act(() => {
      Simulate.change(textarea, { target: { value: 'test' } as Target });
    });
    expect(container).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  test('renders correctly - with listeners', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const onChange = jest.fn();
    act(() => {
      render(
        <UITextarea
          name="test"
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          value="test"
        />,
        container,
      );
    });
    const textarea = document.querySelector('textarea') as HTMLTextAreaElement;
    act(() => {
      Simulate.focus(textarea);
    });
    act(() => {
      Simulate.change(textarea, { target: { value: 'new test' } as Target });
    });
    act(() => {
      Simulate.blur(textarea);
    });
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledWith('test');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('new test');
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledWith('new test');
  });

  test('renders correctly - with listener and debounce', () => {
    const onChange = jest.fn();
    act(() => {
      render(
        <UITextarea
          name="test"
          value="test"
          onChange={onChange}
          debounceTimeout={250}
        />,
        container,
      );
    });
    const textarea = document.querySelector('textarea') as HTMLTextAreaElement;
    act(() => {
      Simulate.change(textarea, { target: { value: 'new test' } as Target });
    });
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('new test');
  });
});
