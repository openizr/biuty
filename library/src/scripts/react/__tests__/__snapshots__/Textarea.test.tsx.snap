// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react/UITextarea renders correctly - autocomplete off 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="off"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - basic 1`] = `
<div>
  <div
    class="ui-textarea ui-textarea--large"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      >
        test
      </textarea>
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - disabled 1`] = `
<div>
  <div
    class="ui-textarea ui-textarea--disabled"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        disabled=""
        id="z8a8d7a5zad4"
        name="test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - readonly 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        readonly=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with cols and rows 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        cols="10"
        id="z8a8d7a5zad4"
        name="test"
        rows="50"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with helper 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      />
    </div>
    <span
      class="ui-textarea__helper"
    >
      test
    </span>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with id 1`] = `
<div>
  <div
    class="ui-textarea"
    id="test"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with label 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <label
      class="ui-textarea__label"
      for="z8a8d7a5zad4"
    >
      <strong>
        test
      </strong>
    </label>
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with listener and debounce 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      >
        new test
      </textarea>
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with listeners 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      >
        new test
      </textarea>
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with maxlength 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        maxlength="10"
        name="test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with placeholder 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        placeholder="test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with transform 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      >
        NEW TEST
      </textarea>
    </div>
  </div>
</div>
`;

exports[`react/UITextarea renders correctly - with value 1`] = `
<div>
  <div
    class="ui-textarea"
  >
    <div
      class="ui-textarea__wrapper"
    >
      <textarea
        autocomplete="on"
        class="ui-textarea__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
      >
        test
      </textarea>
    </div>
  </div>
</div>
`;
